### PR TITLE
Stable Diffusion Enhancements

### PR DESCRIPTION
**Pull request by cmikeh2**
Co-authored-by: Connor Holmes <connorholmes@microsoft.com>
Co-authored-by: Jeff Rasley <jerasley@microsoft.com>

### CODE REVIEW

File: deepspeed/model_implementations/transformers/clip_encoder.py [hunk 1]
```diff
@@ -11,7 +11,13 @@ def __init__(self, enc):
...
         self.device = self.enc.device
         self.dtype = self.enc.dtype
-        self.cuda_graph_created = False
+        self.cuda_graph_created = [False, False]
+        self.static_inputs = [None, None]
+        self.static_kwargs = [None, None]
+        self.static_output = [None, None]
+        self._cuda_graphs = [None, None]
+        self.iter = 0
+        self.enable_cuda_graph = False
```
Comment (RezaYazdaniAminabadi) [inline deepspeed/model_implementations/transformers/clip_encoder.py:20]: let's get this from inference-engine?

File: deepspeed/model_implementations/transformers/clip_encoder.py [hunk 2]
```diff
@@ -27,20 +33,28 @@ def _build_causal_attention_mask(self, bsz, seq_len, dtype):
...
+                self._create_cuda_graph(*inputs, **kwargs)
+                outputs = self._graph_replay(*inputs, **kwargs)
+            self.iter = (self.iter + 1) % 2
+            return outputs
         else:
-            self._create_cuda_graph(*inputs, **kwargs)
-            outputs = self._graph_replay(*inputs, **kwargs)
-        return outputs
+            self.iter += 1
+            if self.iter == 2:
```
Comment (RezaYazdaniAminabadi) [inline deepspeed/model_implementations/transformers/clip_encoder.py:54]: this part of the code is not needed anymore, we can have the enable-cuda-graph enabled from beginning.

File: deepspeed/module_inject/replace_module.py [hunk 3]
```diff
@@ -233,41 +233,46 @@ def transpose(data):
...
-            new_policies = {cross_attention: replace_attn}
+            attention_block = diffusers.models.attention.BasicTransformerBlock
+            new_policies = {
+                cross_attention: replace_attn,
+                attention_block: replace_attn_block,
+            }
         except ImportError:
             new_policies = {}
 
         #replace_transformer_layer(None,
```
Comment (RezaYazdaniAminabadi) [inline deepspeed/module_inject/replace_module.py:254]: do we wanna bring this back if there is no accuracy impact?

**Review by RezaYazdaniAminabadi — APPROVED** [review]

**Review by jeffra — APPROVED** [review]
